tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -12773,6 +12830,30 @@ HLSLMaxRecordsAttr *ValidateMaxRecordsAttributes(Sema &S, Decl *D,
                          A.getAttributeSpellingListIndex());
 }
 
+HLSLWaveSizeAttr *ValidateWaveSizeAttributes(Sema &S, Decl *D,
+                                             const AttributeList &A) {
+  // make sure we are in an appropriate shader model
+  const auto *SM =
+      hlsl::ShaderModel::GetByName(S.getLangOpts().HLSLProfile);
+  if (!SM->IsSM66Plus()) {
+    S.Diag(A.getLoc(), diag::err_hlsl_attribute_in_wrong_shader_model) << "wavesize" << "6.6";
+    return nullptr;
+  }
+
+  // validate that the wavesize argument is a power of 2 between 4 and 128 inclusive
+  HLSLWaveSizeAttr *pAttr = ::new (S.Context)
+      HLSLWaveSizeAttr(A.getRange(), S.Context, ValidateAttributeIntArg(S, A),
+                       A.getAttributeSpellingListIndex());
+
+  unsigned waveSize = pAttr->getSize();
+  if (!DXIL::IsValidWaveSizeValue(waveSize)) {
+    S.Diag(A.getLoc(), diag::err_hlsl_wavesize_size);
```

Comment:
I have a mild preference for keeping the error message dependent on the internally defined constants. It may not be likely in the short term, but it is possible we change those at some point and it seems better design to let it all cascade from those constants.